### PR TITLE
III-5979/bugfix updating place with event

### PR DESCRIPTION
**Pull request by grubolsch**
### Added
Add check that you cannot get a place with a valid event id, or an event with a valid place id.
This could be because the consumer of the API switched up the endpoints of the API.

---
Ticket: https://jira.uitdatabank.be/browse/III-5979/bugfix

### CODE REVIEW

**Comment by simon-debruijn** [comment]
Is it still a WIP? Then you can set it to draft

File: features/Steps/EventSteps.php [hunk 1]
```diff
@@ -58,6 +58,19 @@ public function iGetTheEventAt(string $url): void
         $this->theResponseBodyShouldBeValidJson();
     }
 
+    /**
+     * @Given I fail to get the event at :url
+     */
+    public function iFailToGetTheEventAt(string $url): void
```
Comment (JonasVHG) [inline features/Steps/EventSteps.php:64]: I think the term `FailToGet` is too broad, is it because it does not exist, you do not have permission, it is in the past,...
Since it's a `404` maybe something along the line of `iDoNotFindtheEvent()`

File: features/Steps/PlaceSteps.php [hunk 2]
```diff
@@ -92,6 +92,19 @@ public function iGetThePlaceAt(string $url): void
         $this->theResponseBodyShouldBeValidJson();
     }
 
+    /**
+     * @Given I fail to get the place at :url
+     */
+    public function iFailToGetThePlaceAt(string $url): void
```
Comment (JonasVHG) [inline features/Steps/PlaceSteps.php:98]: Same as above

File: src/Http/CheckTypeOfOfferMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            $offerType = $routeParameters->getOfferType();
+        } catch (RuntimeException $e) {
+            return $handler->handle($request);
+        }
+
+        $offerId = $routeParameters->getOfferId();
+
+        if (OfferType::event()->sameAs($offerType)) {
+            $this->eventRepository->fetch($offerId);
+        } elseif (OfferType::place()->sameAs($offerType)) {
```
Comment (JonasVHG) [inline src/Http/CheckTypeOfOfferMiddleware.php:41]: The elseif isn't necessary.
```suggestion
        }
        if (OfferType::place()->sameAs($offerType)) {
```
has the same logic since $offerType can't be both.

File: app/Authentication/AuthServiceProvider.php [hunk 4]
```diff
@@ -42,6 +43,13 @@ public function register(): void
 
         CurrentUser::configureGodUserIds($container->get('config')['user_permissions']['allow_all']);
 
+        $container->addShared(CheckTypeOfOfferMiddleware::class, function () use ($container): CheckTypeOfOfferMiddleware {
```
Comment (LucWollants) [inline app/Authentication/AuthServiceProvider.php:46]: This new class `CheckTypeOfOfferMiddleware` that is now part of the container also needs to be exposed with `getProvidedServiceNames`.

I also think a better location is `PsrRouterServiceProvider` or `OfferServiceProvider` because this new middleware is not related to authentication.

File: features/Steps/EventSteps.php (hunk 1 above)
Comment (LucWollants) [reply features/Steps/EventSteps.php:64]: I would say there is even no reason to write a new step.

The scenario should be:
- Create a new event and store the UUID
- Get a place with that UUID and check if the error response is 404

Comment (LucWollants) [reply features/Steps/EventSteps.php:64]: Maybe a step like `iGetTheEventAtAndTheResponseIs` could help.

File: features/place/read.feature [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  Background:
+    Given I am using the UDB3 base URL
+    And I am using an UiTID v1 API key of consumer "uitdatabank"
+    And I am authorized as JWT provider v1 user "centraal_beheerder"
+    And I send and accept "application/json"
+
+  @bugfix # https://jira.uitdatabank.be/browse/III-5979
+  Scenario: Try to get a event that actually is a place
+    Given I create a minimal place and save the "url" as "placeUrl"
+    # The line below gets an event which is actually a place - this is on purpose and should fail
```
Comment (LucWollants) [inline features/place/read.feature:12]: When using a step like `iGetTheEventAtAndTheResponseIs` the comment is not needed. Furthermore the test is marked with a bug ticket so all is clear from the ticket.

File: features/place/read.feature [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    Then I fail to get the event at "%{placeUrl}"
+
+  @bugfix # https://jira.uitdatabank.be/browse/III-5979
```
Comment (LucWollants) [inline features/place/read.feature:15]: This test should move to the event features

File: features/Steps/PlaceSteps.php [hunk 7]
```diff
@@ -92,6 +92,32 @@ public function iGetThePlaceAt(string $url): void
...
+        );
+
+        $this->theResponseStatusShouldBe($statusCode);
+        $this->theResponseBodyShouldBeValidJson();
+    }
+
+    /**
+     * @Given I fail to get the place at :url
+     */
+    public function iFailToGetThePlaceAt(string $url): void
```
Comment (LucWollants) [inline features/Steps/PlaceSteps.php:111]: This method is no longer used

File: features/event/read.feature [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  Background:
+    Given I am using the UDB3 base URL
+    And I am using an UiTID v1 API key of consumer "uitdatabank"
+    And I am authorized as JWT provider v1 user "centraal_beheerder"
+    And I send and accept "application/json"
+
+  @bugfix # https://jira.uitdatabank.be/browse/III-5979
+  Scenario: Try to get a event that actually is a place
+    Given I create a minimal place and save the "id" as "placeId"
+    Then I get the event at "http://host.docker.internal:8000/events/%{placeId}" and get response code "404"
```
Comment (LucWollants) [inline features/event/read.feature:12]: The URL should not be hardcoded because this will give issues if we run them in the cloud.
A better approach is to do something like `iGetTheRdfOfEventWithId` and make a method like `iGetTheEventWithIdTheResponseCodeShouldBe`

File: src/Security/PermissionVoterCommandBusSecurity.php [hunk 9]
```diff
@@ -13,7 +13,7 @@ final class PermissionVoterCommandBusSecurity implements CommandBusSecurity
     private PermissionVoter $permissionVoter;
 
     public function __construct(
-        ?string $userId,
+        string $userId,
```
Comment (LucWollants) [inline src/Security/PermissionVoterCommandBusSecurity.php:16]: Is this change needed in the context of this pull request?

File: tests/Http/CheckTypeOfOfferMiddlewareTest.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+        $placeRepository = new InMemoryDocumentRepository();
+        $placeRepository->save(new JsonDocument(self::PLACE_ID, '{}'));
+
+        $eventRepository = new InMemoryDocumentRepository();
+        $eventRepository->save(new JsonDocument(self::EVENT_ID, '{}'));
+
+        $middleware = new CheckTypeOfOfferMiddleware($placeRepository, $eventRepository);
+
+        if ($expectException) {
```
Comment (LucWollants) [inline tests/Http/CheckTypeOfOfferMiddlewareTest.php:40]: The happy paths don't have an assert or expect so you introduced risky tests, see https://github.com/cultuurnet/udb3-backend/actions/runs/7627184604/job/20775262612?pr=1584#step:8:91

**Review by LucWollants — CHANGES_REQUESTED** [review]
This still requires some cleanup work.

File: src/Security/PermissionVoterCommandBusSecurity.php (hunk 9 above)
Comment (grubolsch) [reply src/Security/PermissionVoterCommandBusSecurity.php:16]: No, but i am really sure this is a bug, for 2 reasons:
- It makes no sense that the property declaration and the constructor are different.
- When I did the opposite - make it both nullable, it just failed with other dependencies down the line.

If you want I can cherrypick this into the worlds smallest PR? ;-)

File: tests/Http/CheckTypeOfOfferMiddlewareTest.php (hunk 10 above)
Comment (grubolsch) [reply tests/Http/CheckTypeOfOfferMiddlewareTest.php:40]: Problem is I don't really know what to test here. Before I mocked the EventRepo and placeRepo, and then I had ->expects($this->once()) and the correct repo, so at least I tested that.
I lost that because I now removed the mock.

I can reintroduce the mock if you want?

Comment (LucWollants) [reply tests/Http/CheckTypeOfOfferMiddlewareTest.php:40]: I think there is no need to test the happy paths, those are handled in various other unit tests

File: src/Security/PermissionVoterCommandBusSecurity.php (hunk 9 above)
Comment (LucWollants) [reply src/Security/PermissionVoterCommandBusSecurity.php:16]: But the `getId` method from `CurrentUser` returns `?string` see https://github.com/cultuurnet/udb3-backend/blob/b491c80878447306225bac139da04f6e10deccc3/src/User/CurrentUser.php#L24 so method also has a wrong return type?

File: features/event/read.feature [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  Background:
+    Given I am using the UDB3 base URL
+    And I am using an UiTID v1 API key of consumer "uitdatabank"
+    And I am authorized as JWT provider v1 user "centraal_beheerder"
+    And I send and accept "application/json"
+
+  @bugfix # https://jira.uitdatabank.be/browse/III-5979
+  Scenario: Try to get a event that actually is a place
+    Given I create a minimal place and save the "id" as "placeId"
+    Then I send a GET request to "/events/%{placeId}"
```
Comment (LucWollants) [inline features/event/read.feature:12]:
```suggestion
    When I send a GET request to "/events/%{placeId}"
```

File: features/place/read.feature [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  Background:
+    Given I am using the UDB3 base URL
+    And I am using an UiTID v1 API key of consumer "uitdatabank"
+    And I am authorized as JWT provider v1 user "centraal_beheerder"
+    And I send and accept "application/json"
+
+  @bugfix # https://jira.uitdatabank.be/browse/III-5979
+  Scenario: Try to get a place that actually is an event
+    Given I create a minimal place and save the "url" as "placeUrl"
+    Given I create a minimal permanent event and save the "id" as "eventId"
```
Comment (LucWollants) [inline features/place/read.feature:12]:
```suggestion
    And I create a minimal permanent event and save the "id" as "eventId"
```

File: features/place/read.feature [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    Then I send a GET request to "/places/%{eventId}"
```
Comment (LucWollants) [inline features/place/read.feature:13]:
```suggestion
    When I send a GET request to "/places/%{eventId}"
```

File: tests/Http/CheckTypeOfOfferMiddlewareTest.php (hunk 10 above)
Comment (grubolsch) [reply tests/Http/CheckTypeOfOfferMiddlewareTest.php]: Removed the happy paths

**Review by LucWollants — APPROVED** [review]
